packages/app/android/src/main/java/io/invertase/firebase/common/UniversalFirebaseModule.java

```diff
@@ -68,12 +70,21 @@ private ExecutorService getExecutor(boolean isTransactional) {
 
   private ExecutorService getNewExecutor(boolean isTransactional) {
     if (isTransactional == true) {
-      return new ThreadPoolExecutor(0, 1, 60L, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>());
+      return Executors.newSingleThreadExecutor();
     } else {
-      return new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60L, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());
+      ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(0, MAXIMUM_POOL_SIZE, KEEP_ALIVE_SECONDS, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());
+      threadPoolExecutor.setRejectedExecutionHandler(executeInFallback);
+      return threadPoolExecutor;
     }
   }
 
+  private RejectedExecutionHandler executeInFallback = new RejectedExecutionHandler() {
```

Comment:
Is the RejectedExecutionHandler intended to solve a problem where the multi-thread executor was in an "all threads busy" state, and rejected tasks that would have had to wait but otherwise executed just fine in the prior SingleThreadExecutor?

If so, I like this as a fallback strategy, however, it means that if the multi-thread executor shuts down first, all of it's tasks will spill over to the SingleThreadExecutor, only to immediately be shutdown there:

https://developer.android.com/reference/java/util/concurrent/RejectedExecutionHandler#rejectedExecution(java.lang.Runnable,%20java.util.concurrent.ThreadPoolExecutor)

> Method that may be invoked by a ThreadPoolExecutor when execute cannot accept a task. This may occur when no more threads or queue slots are available because their bounds would be exceeded, **or upon shutdown of the Executor.**

(emphasis mine)

I think it is possible to stop a cascade of shutdown rejection from executor to executor if you gate task resubmission to the SingleThreadExecutor on a check of the rejecting executor's isTerminating/isShutdown/isTerminated status

https://developer.android.com/reference/java/util/concurrent/ThreadPoolExecutor#isTerminating() (and the other two)

Reply:
> it means that if the multi-thread executor shuts down first, all of it's tasks will spill over to the SingleThreadExecutor, only to immediately be shutdown there:

Yeah, it's true! Your suggestion is correct and looks straightforward.